File: clients/projects/office/src/app/modules/core/services/multitenancy/billing.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { DateRangeQuerySearchFilter } from '@xyz/office/modules/shared/modules/query-search-filter';

import { BillingInvoice } from '../../entities';
import { Page, PageRequest } from '../../models';
import { EnvironmentService } from '../environment.service';

@Injectable({
  providedIn: 'root'
})
export class BillingService {
  private readonly _endpointSlug: string = 'admin/billing';

  constructor(
    private _environmentService: EnvironmentService,
    private _http: HttpClient
  ) { }

  public searchBillingInvoices(filter: DateRangeQuerySearchFilter, pageRequest: PageRequest): Observable<Page<BillingInvoice>> {
    const queryParms: {[key: string]: string} = {
      query: filter?.query || '',
      startDate: filter?.startDate || '',
      endDate: filter?.endDate || '',
      size: pageRequest.size.toString(),
      index: pageRequest.index.toString(),
      column: pageRequest?.sort?.column?.toString() || '',
      direction: pageRequest?.sort?.direction?.toString() || ''
    };
    return this._http.get<Page<BillingInvoice>>(
      `${this._environmentService.getBaseApiUrl()}/${this._endpointSlug}/invoices/search`,
      { params: queryParms }
    );
  }
}

File: server/src/Xyz.Infrastructure/Services/Multitenancy/BillingService.cs
using Microsoft.Extensions.Logging;
using Microsoft.EntityFrameworkCore;

using Xyz.Core.Models.Paging;
using Xyz.Core.Models.SearchFilters;
using Xyz.Core.Models.Multitenancy;
using Xyz.Core.Interfaces.Multitenancy;

using Xyz.Core.Entities.Multitenancy;

using Xyz.Multitenancy.Data;

namespace Xyz.Infrastructure.Services.Multitenancy
{
    public class BillingService : IBillingService
    {
        private readonly ILogger<BillingService> _logger;
        private readonly MultitenancyDbContext _multitenancyDbContext;

        public BillingService(
            ILogger<BillingService> logger,
            MultitenancyDbContext multitenancyDbContext)
        {
            this._logger = logger;
            this._multitenancyDbContext = multitenancyDbContext;
        }

        public async Task<Page<BillingInvoiceDto>> SearchBillingInvoices(
            string tenantId, 
            DateRangeQuerySearchFilter filter, 
            PageRequest pageRequest)
        {
            try
            {
                IQueryable<BillingInvoice> query = this._multitenancyDbContext.BillingInvoices
                    .Include(bi => bi.Tenant);

                if (filter?.Query != null)
                {
                    var queryTerm = filter.Query.Trim().ToLower();
                    query = query.Where(bi => 
                        bi.AmountPaid.ToString().Contains(queryTerm)
                            || nameof(bi.Status).ToString().ToLower().Contains(queryTerm));
                }

                if (filter?.StartDate != null && filter?.EndDate != null)
                {
                    query = query
                        .Where(bi => bi.TransactionDate.Date >= filter.StartDate.Value.Date)
                        .Where(bi => bi.TransactionDate.Date <= filter.EndDate.Value.AddDays(1).Date);
                }
                else if (filter?.StartDate != null)
                {
                    query = query
                        .Where(bi => bi.TransactionDate.Date >= filter.StartDate.Value.Date);
                }
                else if (filter?.EndDate != null)
                {
                    query = query
                        .Where(bi => bi.TransactionDate.Date <= filter.EndDate.Value.AddDays(1).Date);
                }

                if (pageRequest.Sort != null)
                {
                    switch (pageRequest.Sort.Column)
                    {
                        case "transactionDate":
                            query = pageRequest.Sort.Direction == SortDirection.Ascend
                                ? query.OrderBy(bi => bi.TransactionDate)
                                : query.OrderByDescending(bi => bi.TransactionDate);
                            break;
                        case "amountPaid":
                            query = pageRequest.Sort.Direction == SortDirection.Ascend
                                ? query.OrderBy(bi => bi.AmountPaid)
                                : query.OrderByDescending(bi => bi.AmountPaid);
                            break;
                        case "status":
                            query = pageRequest.Sort.Direction == SortDirection.Ascend
                                ? query.OrderBy(bi => bi.Status)
                                : query.OrderByDescending(bi => bi.Status);
                            break;
                        default:
                            query = query.OrderByDescending(t => t.TransactionDate);
                            break;
                    }
                }
                    
                var billingInvoicesSource = query
                    .Select(bi => bi.ToDto());

                return await Page<BillingInvoiceDto>.From(billingInvoicesSource, pageRequest);
            }
            catch (Exception ex)
            {
                var errorMessage = "Error searching billing invoices!";
                this._logger.LogError(errorMessage, new { Exception = ex });
                throw;
            }
        }
    }
}